pytorch_lightning/trainer/progress.py

```diff
@@ -0,0 +1,114 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+
+
+@dataclass
+class LoopProgress:
+    """Basic dataclass to track loop progress during execution.
+
+    The general structure is to count totals across epochs as well as intra-epoch counts.
+    These counters are local to a trainer rank.
+    By default, they are not globally synced across all ranks.
+
+    Args:
+        epochs_finished_total: Number of passes through the dataset.
+            This is intended to be incremented after `on_*_epoch_end` completes.
+        batches_read_total: Number of batches loaded through the dataloader.
+        batches_started_total: Number of batches started processing.
+            This is intended to be incremented after `on_*_batch_start` completes.
+        batches_processed_total: Number of batches processed.
+            This is intended to be incremented after `*_step` runs.
+        batches_finished_total: Number of batches finished.
+            This is intended to be incremented after `on_*_batch_end` runs.
+
+        batches_read_epoch: Number of batches loaded through the dataloader within the current epoch.
+        batches_started_epoch: Number of batches started processing within the current epoch.
+            This is intended to be incremented after `on_*_batch_start` completes.
+        batches_processed_epoch: Number of batches processed within the current epoch.
+            This is intended to be incremented after `*_step` runs.
+        batches_finished_epoch: Number of batches finished within the current epoch.
+            This is intended to be incremented after `on_*_batch_end` runs.
+    """
```

Comment:
discussed w/ @carmocca and @awaelchli 
we will split this up into a smaller class:
```
@dataclass
class Progress:
    ready
    started
    processed
    completed
```
then the Loop Progress can be:
```
@dataclass
class LoopProgress:
    total = Progress()
    epoch = Progress()
```

we can add a separate optimization progress. @carmocca could you share the other notes from our sync here?

Reply:
There you go:

<details>

```python
class TrainerState:
    running_stage: RunningStage
    fn: TrainerFn
    status: TrainerStatus

    loops: LoopProgress
    optimization: OptimizationProgress

class LoopProgress:
    train: TrainLoopProgress
    val: LoopProgress
    test: LoopProgress
    predict: LoopProgress

# passed by the Trainer
class TrainerContext  # immutable for model
    ranks...
    distributed_type
    precision

# lives in the Model
class ModelProgress
    epochs: int
    batches: int

class TrainLoopProgress  # not mutable
    epoch: EpochProgress
    batch: BatchProgress

class OptimizationProgress  # modified by the model in case of manual
    # issue of multilple opt -> assume all optimizers at first
    optimizer_steps: int
    scheduler_steps: int

class Progress:
    ready: int
    started: int
    processed: int
    completed: int

class BatchProgress:
    total = Progress()
    current = Progress()

class EpochProgress:
    total = Progress()
    current = Progress()
```

</details>